### PR TITLE
|SBOM] Use in-memory overlayfs to scan container images

### PR DESCRIPTION
**Pull request by lebauce**
### What does this PR do?

When the container runtime is using overlayfs as filesystem, use a virtual
in memory overlayfs to scan the image instead of relying on the snapshotter.

### Motivation

Using the snapshotter requires having the CAP_SYS_ADMIN capability while this method
"only" needs
 uid 0.

### CODE REVIEW

File: pkg/util/trivy/trivy.go [hunk 1]
```diff
@@ -244,6 +246,29 @@ func (c *Collector) ScanDockerImage(ctx context.Context, imgMeta *workloadmeta.C
...
+		}
+
+		fs := NewFS(layers)
+		report, err := c.scanFilesystem(ctx, fs, ".", nil, scanOptions)
+		if err != nil {
+			return nil, err
+		}
+
+		if report, ok := report.(*Report); ok {
+			report.Report.Metadata.ImageID = imgMeta.ID
```
Comment (AliDatadog) [inline pkg/util/trivy/trivy.go:266]: Do we still need to populate the metadata field ?

File: pkg/util/trivy/trivy.go [hunk 2]
```diff
@@ -244,6 +246,29 @@ func (c *Collector) ScanDockerImage(ctx context.Context, imgMeta *workloadmeta.C
 		return nil, fmt.Errorf("unable to convert docker image, err: %w", err)
 	}
 
+	if fanalImage.inspect.GraphDriver.Name == "overlay2" {
```
Comment (AliDatadog) [inline pkg/util/trivy/trivy.go:249]: I think we would need the same logic for containerd.

File: pkg/sbom/collectors/host/host.go [hunk 3]
```diff
@@ -27,30 +28,31 @@ import (
 // not trigger multiple scans at the same time unlike for container-images.
 const channelSize = 1
 
-// scanRequest defines a scan request. This struct should be
+// ScanRequest defines a scan request. This struct should be
 // hashable to be pushed in the work queue for processing.
-type scanRequest struct {
-	path string
+type ScanRequest struct {
```
Comment (AliDatadog) [inline pkg/sbom/collectors/host/host.go:33]: Can we keep this private ?

File: pkg/util/trivy/trivy.go [hunk 4]
```diff
@@ -292,24 +327,34 @@ func (c *Collector) ScanContainerdImageFromFilesystem(ctx context.Context, imgMe
...
 
 	trivyReport, err := c.scan(ctx, fsArtifact, applier.NewApplier(cache), imgMeta, cache)
 	if err != nil {
+		if imgMeta != nil {
+			return nil, fmt.Errorf("unable to marshal report to sbom format for image %s, err: %w", imgMeta.ID, err)
+		}
 		return nil, fmt.Errorf("unable to marshal report to sbom format, err: %w", err)
 	}
 
+	log.Infof("Found OS: %+v", trivyReport.Metadata.OS)
```
Comment (AliDatadog) [inline pkg/util/trivy/trivy.go:351]: I think it could be a debug log

File: pkg/util/trivy/trivy.go [hunk 5]
```diff
@@ -244,9 +248,36 @@ func (c *Collector) ScanDockerImage(ctx context.Context, imgMeta *workloadmeta.C
...
 
+	if c.config.overlayFSSupport && fanalImage.inspect.GraphDriver.Name == "overlay2" {
+		return c.scanOverlayFS(ctx, fanalImage, scanOptions)
+	}
+
 	return c.scanImage(ctx, fanalImage, imgMeta, scanOptions)
 }
 
+func (c *Collector) scanOverlayFS(ctx context.Context, fanalImage *image, scanOptions sbom.ScanOptions) (sbom.Report, error) {
+	if !c.config.overlayFSSupport || fanalImage.inspect.GraphDriver.Name != "overlay2" {
```
Comment (AliDatadog) [inline pkg/util/trivy/trivy.go:259]: I think we already do this check before calling the function

File: pkg/util/trivy/trivy.go [hunk 6]
```diff
@@ -244,9 +248,36 @@ func (c *Collector) ScanDockerImage(ctx context.Context, imgMeta *workloadmeta.C
...
+	if layerDirs, ok := fanalImage.inspect.GraphDriver.Data["LowerDir"]; ok {
+		layers = append(layers, strings.Split(layerDirs, ":")...)
+	}
+
+	if layerDirs, ok := fanalImage.inspect.GraphDriver.Data["UpperDir"]; ok {
+		layers = append(layers, strings.Split(layerDirs, ":")...)
+	}
+
+	fs := NewFS(layers)
+	report, err := c.scanFilesystem(ctx, fs, ".", nil, scanOptions)
```
Comment (AliDatadog) [inline pkg/util/trivy/trivy.go:273]: Do we need to request a lease to avoid deletion while we're scanning the FS ?

**Review by AliDatadog — APPROVED** [review]

**Review by ogaca-dd — APPROVED** [review]
LGTM for files owned by Agent shared component.

**Review by YoannGh — APPROVED** [review]

**Pull request merged**